gusto/physics.py

```diff
@@ -0,0 +1,120 @@
+from abc import ABCMeta, abstractmethod
+from firedrake import exp, Function, project, conditional, interpolate
+
+
+class Physics(object):
+    """
+    Base class for physics processes for Gusto.
+
+    :arg state: :class:`.State` object.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, state):
+        self.state = state
+
+    @abstractmethod
+    def apply(self):
+        """
+        Function computes the value of specific
+        fields at the next time step.
+        """
+        pass
+
+
+class Condensation(Physics):
+    """
+    The process of condensation of water vapour
+    into liquid water and evaporation of liquid
+    water into water vapour, with the associated
+    latent heat changes.
+
+    :arg state: :class:`.State.` object.
+    """
+
+    def __init__(self, state):
+        super(Condensation, self).__init__(state)
+
+        # obtain our fields
+        self.theta = state.fields('theta')
+        self.water_v = state.fields('water_v')
+        self.water_c = state.fields('water_c')
+        self.rho = state.fields('rho')
+
+        # declare function space
+        V = self.theta.function_space()
+
+        # make functions for fields at next time step
+        self.water_v_new = Function(V)
+        self.water_c_new = Function(V)
+        self.theta_new = Function(V)
+
+    def update_fields(self):
+
+        param = self.state.parameters
+
+        # define some parameters as attributes
+        dt = self.state.timestepping.dt
+        R_d = param.R_d
+        p_0 = param.p_0
+        kappa = param.kappa
+        cp = param.cp
+        cv = param.cv
+        c_pv = param.c_pv
+        c_pl = param.c_pl
+        c_vv = param.c_vv
+        R_v = param.R_v
+        L_v0 = param.L_v0
+        T_0 = param.T_0
+        w_sat1 = param.w_sat1
+        w_sat2 = param.w_sat2
+        w_sat3 = param.w_sat3
+        w_sat4 = param.w_sat4
+
+        # declare function space
+        V = self.theta.function_space()
+
+        # make useful fields
+        Pi = ((R_d * project(self.rho, V) * self.theta / p_0)
```

Comment:
I tried that at first, but I was getting an error like this: `ValueError: Mismatching indexed function spaces`. Since I didn't really know what I was doing I just tried this, which seemed to work...